File: services/api-gateway/src/events.js
const { connect, StringCodec } = require('nats');
const { v4: uuidv4 } = require('uuid');

let nc = null;

const BACKOFF_MS = [1000, 2000, 4000, 8000, 16000];

/**
 * Connect to NATS with exponential backoff (5 retries after initial attempt).
 * Does not throw — logs warnings if NATS stays unavailable.
 */
async function connectNATS() {
  const url = process.env.NATS_URL;
  if (!url) {
    console.warn('NATS_URL not set; NATS disabled');
    return;
  }

  let lastErr;
  for (let attempt = 0; attempt <= BACKOFF_MS.length; attempt += 1) {
    try {
      nc = await connect({ servers: url });
      console.info('NATS connected');
      return;
    } catch (err) {
      lastErr = err;
      console.warn(`NATS connect attempt ${attempt + 1} failed:`, err.message);
      if (attempt < BACKOFF_MS.length) {
        await new Promise((resolve) => setTimeout(resolve, BACKOFF_MS[attempt]));
      }
    }
  }
  console.warn('NATS unavailable after retries; continuing without NATS:', lastErr?.message);
}

/**
 * Publish event with contract shape. No-op if NATS is down (service keeps running).
 */
function publishEvent(subject, data) {
  if (!nc) {
    console.warn('publishEvent skipped (no NATS):', subject);
    return;
  }
  const sc = StringCodec();
  const payload = {
    event_id: uuidv4(),
    subject,
    service: 'api-gateway',
    timestamp: new Date().toISOString(),
    data,
  };
  try {
    nc.publish(subject, sc.encode(JSON.stringify(payload)));
  } catch (err) {
    console.warn('publishEvent failed:', err.message);
  }
}

async function drainNATS() {
  if (!nc) return;
  try {
    await nc.drain();
  } catch (err) {
    console.warn('NATS drain:', err.message);
  }
  nc = null;
}

module.exports = { connectNATS, publishEvent, drainNATS };

File: services/auth-service/src/events.py
"""NATS — connect with retry; publish BaseEvent JSON (PRD event contract)."""

from __future__ import annotations

import asyncio
import json
import logging
import uuid
from datetime import UTC, datetime
from typing import Any

import nats
from nats.aio.client import Client as NatsClient

from src.settings import get_settings

logger = logging.getLogger(__name__)

BACKOFF_SEC = [1, 2, 4, 8, 16]

_nc: NatsClient | None = None


async def connect_nats() -> None:
    global _nc
    s = get_settings()
    if not s.nats_url:
        logger.warning("NATS_URL not set; NATS disabled")
        return

    max_attempts = len(BACKOFF_SEC) + 1
    last_err: Exception | None = None
    for attempt in range(max_attempts):
        try:
            _nc = await nats.connect(servers=[s.nats_url])
            logger.info("NATS connected (auth-service)")
            return
        except Exception as e:
            last_err = e
            logger.warning(
                "NATS connect attempt %s/%s failed: %s",
                attempt + 1,
                max_attempts,
                e,
            )
            if attempt < len(BACKOFF_SEC):
                await asyncio.sleep(BACKOFF_SEC[attempt])

    logger.warning(
        "NATS unavailable after retries; continuing without NATS: %s",
        last_err,
    )


async def publish(subject: str, data: dict[str, Any]) -> None:
    if _nc is None:
        logger.warning("publish skipped (no NATS): %s", subject)
        return
    ts = datetime.now(UTC).isoformat().replace("+00:00", "Z")
    payload = {
        "event_id": str(uuid.uuid4()),
        "subject": subject,
        "service": "auth-service",
        "timestamp": ts,
        "data": data,
    }
    try:
        await _nc.publish(subject, json.dumps(payload).encode("utf-8"))
    except Exception as e:
        logger.warning("publish failed: %s", e)


async def drain_nats() -> None:
    global _nc
    if _nc is None:
        return
    try:
        await _nc.drain()
    except Exception as e:
        logger.warning("NATS drain: %s", e)
    _nc = None
